Firestore/core/src/firebase/firestore/core/transaction_runner.mm

```diff
@@ -0,0 +1,114 @@
+
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/core/transaction_runner.h"
+#include "Firestore/core/src/firebase/firestore/remote/exponential_backoff.h"
+
+#include "absl/algorithm/container.h"
+
+using firebase::firestore::core::TransactionResultCallback;
+using firebase::firestore::core::TransactionUpdateCallback;
+using firebase::firestore::remote::Datastore;
+using firebase::firestore::remote::RemoteStore;
+using firebase::firestore::util::AsyncQueue;
+using firebase::firestore::util::TimerId;
+using firebase::firestore::util::Status;
+
+namespace firebase {
+namespace firestore {
+namespace core {
+TransactionRunner::TransactionRunner(const std::shared_ptr<AsyncQueue>& queue,
+                                     RemoteStore* remote_store,
+                                     TransactionUpdateCallback update_callback,
+                                     TransactionResultCallback result_callback)
+    : queue_{queue},
+      remote_store_{remote_store},
+      update_callback_{update_callback},
+      result_callback_{result_callback},
+      backoff_{queue_, TimerId::RetryTransaction} {
+  retries_left_ = kRetryCount;
+}
+
+void TransactionRunner::Run() {
+  queue_->VerifyIsCurrentQueue();
+
+  auto shared_this = this->shared_from_this();
+  backoff_.BackoffAndRun([shared_this] {
+    std::shared_ptr<Transaction> transaction =
+        shared_this->remote_store_->CreateTransaction();
+    shared_this->update_callback_(
+        transaction,
+        [transaction, shared_this](util::StatusOr<absl::any> maybe_result) {
+          shared_this->queue_->Enqueue(
+              [transaction, shared_this, maybe_result] {
+                shared_this->RunTransaction(transaction, maybe_result);
+              });
+        });
+  });
+}
+
+void TransactionRunner::RunTransaction(
+    const std::shared_ptr<Transaction> transaction,
+    const util::StatusOr<absl::any> maybe_result) {
+  if (!maybe_result.ok()) {
+    TransactionRunner::HandleTransactionError(transaction,
+                                              maybe_result.status());
+  } else {
+    auto shared_this = this->shared_from_this();
+    transaction->Commit(
+        [shared_this, transaction, maybe_result](Status status) {
+          shared_this->CheckCommitResult(transaction, status, maybe_result);
+        });
+  }
+}
+
+void TransactionRunner::CheckCommitResult(
+    const std::shared_ptr<Transaction> transaction,
+    Status status,
+    const util::StatusOr<absl::any> maybe_result) {
+  if (status.ok()) {
+    result_callback_(std::move(maybe_result));
+  } else {
+    TransactionRunner::HandleTransactionError(transaction, status);
+  }
+}
+
+void TransactionRunner::HandleTransactionError(
+    const std::shared_ptr<Transaction> transaction, Status status) {
+  if (retries_left_ > 0 &&
+      TransactionRunner::IsRetryableTransactionError(status) &&
+      !transaction->IsPermanentlyFailed()) {
+    retries_left_ -= 1;
+    TransactionRunner::Run();
```

Comment:
Is there a reason this isn't just `Run()`?

Reply:
No reason at all -- when the code finally built successfully, I was too happy to notice.

I've changed the other ones as well.